File: treeTraversal.js
class TreeNode {
    constructor(key) {
        this.key = key;
        this.left = null;   
        this.right = null;  
    }
}

class BinaryTree {
    constructor() {
        this.root = null;
    }

    insert(key) {
        if (this.root === null) {
            this.root = new TreeNode(key);
        } else {
            this._insertRecur(this.root, key);
        }
    }

    _insertRecur(node, key) {
        if (key < node.key) {
            if (node.left === null) {
                node.left = new TreeNode(key);
            } else {
                this._insertRecur(node.left, key);
            }
        } else if (key > node.key) {
            if (node.right === null) {
                node.right = new TreeNode(key);
            } else {
                this._insertRecur(node.right, key);
            }
        }
    }

    inOrderTrav() {
        let result = [];
        this._inOrderTravRecur(this.root, result);
        console.log(result.join(" "));
    }

    _inOrderTravRecur(node, result) {
        if (node) {
            this._inOrderTravRecur(node.left, result);
            result.push(node.key);
            this._inOrderTravRecur(node.right, result);
        }
    }

    preOrderTrav() {
        let result = [];
        this._preOrderTravRecur(this.root, result);
        console.log(result.join(" "));
    }

    _preOrderTravRecur(node, result) {
        if (node) {
            result.push(node.key);
            this._preOrderTravRecur(node.left, result);
            this._preOrderTravRecur(node.right, result);
        }
    }

    postOrderTrav() {
        let result = [];
        this._postOrderTravRecur(this.root, result);
        console.log(result.join(" "));
    }

    _postOrderTravRecur(node, result) {
        if (node) {
            this._postOrderTravRecur(node.left, result);
            this._postOrderTravRecur(node.right, result);
            result.push(node.key);
        }
    }

    logTree() {
        this._logTreeRecur(this.root, 0);
    }

    _logTreeRecur(node, depth) {
        if (node === null) {
            return;
        }
        this._logTreeRecur(node.right, depth + 1);
        console.log("   ".repeat(depth) + String(node.key));
        this._logTreeRecur(node.left, depth + 1)
    }
}

const tree = new BinaryTree();

// Insert keys
const keys = [50, 30, 70, 40, 20, 60, 80];
keys.forEach(key => tree.insert(key));

// In-order traversal
console.log('In-order traversal:');
tree.inOrderTrav();

// Pre-order traversal
console.log('Pre-order traversal:');
tree.preOrderTrav();


// Post-order traversal
console.log('Post-order traversal:');
tree.postOrderTrav();

// Log tree structure
console.log('Binary tree structure:');
tree.logTree();

File: tree_traversal.py
class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if self.root is None:
            self.root = TreeNode(key)
        else:
            self._insert_recursive(self.root, key)

    def _insert_recursive(self, node, key):
        if key < node.key:
            if node.left is None:
                node.left = TreeNode(key)
            else:
                self._insert_recursive(node.left, key)
        elif key > node.key:
            if node.right is None:
                node.right = TreeNode(key)
            else:
                self._insert_recursive(node.right, key)

    def in_order_trav(self):
        self._in_order_trav(self.root)
        print()

    def _in_order_trav(self, node):
        if node:
            self._in_order_trav(node.left)
            print(node.key, end=" ")
            self._in_order_trav(node.right)

    def pre_order_trav(self):
        self._pre_order_trav_recur(self.root)
        print()

    def _pre_order_trav_recur(self, node):
        if node:
            print(node.key, end=" ")
            self._pre_order_trav_recur(node.left)
            self._pre_order_trav_recur(node.right)

    def post_order_trav(self):
        self._post_order_trav_recur(self.root)
        print()

    def _post_order_trav_recur(self, node):
        if node:
            self._post_order_trav_recur(node.left)
            self._post_order_trav_recur(node.right)
            print(node.key, end=" ")

    def print_tree(self):
        self._print_tree_recur(self.root, 0)

    def _print_tree_recur(self, node, depth):
        if node is None:
            return
        self._print_tree_recur(node.right, depth + 1)
        print("   " * depth + str((node.key)))
        self._print_tree_recur(node.left, depth + 1)



if __name__ == "__main__":
    tree = BinaryTree()

    # Insert keys
    keys = [50, 30, 70, 40, 20, 60, 80]
    for key in keys:
        tree.insert(key)

    # In-order traversal
    print('In-order traversal:')
    tree.in_order_trav()

    # Pre-order traversal
    print('Pre-order traversal:')
    tree.pre_order_trav()

    # Post-order traversal
    print('Post-order traversal:')
    tree.post_order_trav()

    # Print Tree Structure
    print('Binary tree structure:')
    tree.print_tree()
